File: packages/apidash_mcp_core/src/codegen.ts
/**
 * Code generator utility for APIDash MCP
 * Generates code snippets for HTTP requests in various languages
 */

export const SUPPORTED_GENERATORS = [
  "curl", "python-requests", "javascript-fetch", "javascript-axios",
  "nodejs-fetch", "dart-http", "go-http", "java-http",
  "kotlin-okhttp", "php-curl", "ruby-net", "rust-reqwest"
] as const;

export type GeneratorId = typeof SUPPORTED_GENERATORS[number];

export interface CodeGenInput {
  method: string;
  url: string;
  headers?: Record<string, string>;
  body?: string;
  contentType?: string;
}

function headersToEntries(headers?: Record<string, string>): [string, string][] {
  return Object.entries(headers ?? {});
}

function sanitizeUrl(url: string): string {
  return url.replace(/'/g, "\\'");
}

export function generateCurl(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = [`curl -X ${method} '${sanitizeUrl(url)}'`];
  for (const [k, v] of headersToEntries(headers)) {
    lines.push(`  -H '${k}: ${v}'`);
  }
  if (body) {
    lines.push(`  --data '${body.replace(/'/g, "\\'")}'`);
  }
  return lines.join(' \\\n');
}

export function generatePythonRequests(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = ['import requests', ''];
  lines.push(`url = "${url}"`);

  if (headers && Object.keys(headers).length > 0) {
    lines.push('headers = {');
    for (const [k, v] of headersToEntries(headers)) {
      lines.push(`    "${k}": "${v}",`);
    }
    lines.push('}');
  } else {
    lines.push('headers = {}');
  }

  if (body) {
    lines.push('');
    lines.push(`payload = '''${body}'''`);
    lines.push('');
    lines.push(`response = requests.${method.toLowerCase()}(url, headers=headers, data=payload)`);
  } else {
    lines.push('');
    lines.push(`response = requests.${method.toLowerCase()}(url, headers=headers)`);
  }

  lines.push('print(response.status_code)');
  lines.push('print(response.json())');
  return lines.join('\n');
}

export function generateJavaScriptFetch(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = [];
  lines.push(`const response = await fetch("${url}", {`);
  lines.push(`  method: "${method}",`);

  if (headers && Object.keys(headers).length > 0) {
    lines.push('  headers: {');
    for (const [k, v] of headersToEntries(headers)) {
      lines.push(`    "${k}": "${v}",`);
    }
    lines.push('  },');
  }

  if (body) {
    const escaped = body.replace(/\\/g, '\\\\').replace(/`/g, '\\`').replace(/\${/g, '\\${');
    lines.push(`  body: \`${escaped}\`,`);
  }

  lines.push('});');
  lines.push('');
  lines.push('const data = await response.json();');
  lines.push('console.log(data);');
  return lines.join('\n');
}

export function generateJavaScriptAxios(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = ['import axios from "axios";', ''];
  lines.push('const config = {');
  lines.push(`  method: "${method.toLowerCase()}",`);
  lines.push(`  url: "${url}",`);

  if (headers && Object.keys(headers).length > 0) {
    lines.push('  headers: {');
    for (const [k, v] of headersToEntries(headers)) {
      lines.push(`    "${k}": "${v}",`);
    }
    lines.push('  },');
  }

  if (body) {
    lines.push(`  data: ${body},`);
  }

  lines.push('};');
  lines.push('');
  lines.push('const { data } = await axios(config);');
  lines.push('console.log(data);');
  return lines.join('\n');
}

export function generateNodeFetch(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = ['const fetch = require("node-fetch");', ''];
  lines.push('async function run() {');
  lines.push(`  const response = await fetch("${url}", {`);
  lines.push(`    method: "${method}",`);
  if (headers && Object.keys(headers).length > 0) {
    lines.push('    headers: {');
    for (const [k, v] of headersToEntries(headers)) {
      lines.push(`      "${k}": "${v}",`);
    }
    lines.push('    },');
  }
  if (body) {
    lines.push(`    body: JSON.stringify(${body}),`);
  }
  lines.push('  });');
  lines.push('  const data = await response.json();');
  lines.push('  console.log(data);');
  lines.push('}');
  lines.push('run();');
  return lines.join('\n');
}

export function generateDartHttp(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = ["import 'package:http/http.dart' as http;", ''];
  lines.push('void main() async {');
  lines.push(`  final uri = Uri.parse('${url}');`);
  if (headers && Object.keys(headers).length > 0) {
    lines.push('  final headers = {');
    for (const [k, v] of headersToEntries(headers)) {
      lines.push(`    '${k}': '${v}',`);
    }
    lines.push('  };');
  }
  if (body) {
    lines.push(`  final body = '''${body}''';`);
    lines.push(`  final response = await http.${method.toLowerCase()}(uri, headers: ${headers ? 'headers' : '{}'}, body: body);`);
  } else {
    lines.push(`  final response = await http.${method.toLowerCase()}(uri${headers ? ', headers: headers' : ''});`);
  }
  lines.push('  print(response.statusCode);');
  lines.push('  print(response.body);');
  lines.push('}');
  return lines.join('\n');
}

export function generateGo(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = [
    'package main', '',
    'import (', '  "fmt"', '  "io"', '  "net/http"', '  "strings"', ')', '',
    'func main() {',
  ];
  if (body) {
    lines.push(`  payload := strings.NewReader(\`${body}\`)`);
    lines.push(`  req, _ := http.NewRequest("${method}", "${url}", payload)`);
  } else {
    lines.push(`  req, _ := http.NewRequest("${method}", "${url}", nil)`);
  }
  for (const [k, v] of headersToEntries(headers)) {
    lines.push(`  req.Header.Add("${k}", "${v}")`);
  }
  lines.push('  client := &http.Client{}');
  lines.push('  resp, _ := client.Do(req)');
  lines.push('  defer resp.Body.Close()');
  lines.push('  body, _ := io.ReadAll(resp.Body)');
  lines.push('  fmt.Println(resp.StatusCode)');
  lines.push('  fmt.Println(string(body))');
  lines.push('}');
  return lines.join('\n');
}

export function generateJava(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = [
    'import java.net.URI;',
    'import java.net.http.HttpClient;',
    'import java.net.http.HttpRequest;',
    'import java.net.http.HttpResponse;',
    '',
    'public class ApiRequest {',
    '  public static void main(String[] args) throws Exception {',
    '    var client = HttpClient.newHttpClient();',
  ];
  lines.push(`    var request = HttpRequest.newBuilder()`);
  lines.push(`        .uri(URI.create("${url}"))`);
  for (const [k, v] of headersToEntries(headers)) {
    lines.push(`        .header("${k}", "${v}")`);
  }
  if (body) {
    lines.push(`        .${method}(HttpRequest.BodyPublishers.ofString("""${body}"""))`);
  } else {
    lines.push(`        .${method}(HttpRequest.BodyPublishers.noBody())`);
  }
  lines.push('        .build();');
  lines.push('    var response = client.send(request, HttpResponse.BodyHandlers.ofString());');
  lines.push('    System.out.println(response.statusCode());');
  lines.push('    System.out.println(response.body());');
  lines.push('  }');
  lines.push('}');
  return lines.join('\n');
}

export function generateKotlin(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = ['import okhttp3.OkHttpClient', 'import okhttp3.Request', ''];
  if (body) {
    lines.push('import okhttp3.RequestBody.Companion.toRequestBody');
    lines.push('import okhttp3.MediaType.Companion.toMediaType');
    lines.push('');
  }
  lines.push('val client = OkHttpClient()');
  if (body) {
    lines.push(`val body = """${body}""".toRequestBody("application/json".toMediaType())`);
  }
  lines.push('val request = Request.Builder()');
  lines.push(`    .url("${url}")`);
  for (const [k, v] of headersToEntries(headers)) {
    lines.push(`    .addHeader("${k}", "${v}")`);
  }
  if (body) {
    lines.push(`    .${method.toLowerCase()}(body)`);
  } else {
    lines.push(`    .${method.toLowerCase()}()`);
  }
  lines.push('    .build()');
  lines.push('');
  lines.push('val response = client.newCall(request).execute()');
  lines.push('println(response.code)');
  lines.push('println(response.body?.string())');
  return lines.join('\n');
}

export function generatePhp(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = ['<?php', ''];
  lines.push(`$ch = curl_init("${url}");`);
  lines.push('curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);');
  lines.push(`curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "${method}");`);
  if (headers && Object.keys(headers).length > 0) {
    lines.push('curl_setopt($ch, CURLOPT_HTTPHEADER, [');
    for (const [k, v] of headersToEntries(headers)) {
      lines.push(`  "${k}: ${v}",`);
    }
    lines.push(']);');
  }
  if (body) {
    lines.push(`curl_setopt($ch, CURLOPT_POSTFIELDS, '${body.replace(/'/g, "\\'")}');`);
  }
  lines.push('');
  lines.push('$response = curl_exec($ch);');
  lines.push('$statusCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);');
  lines.push('curl_close($ch);');
  lines.push('');
  lines.push('echo $statusCode . "\\n";');
  lines.push('echo $response . "\\n";');
  return lines.join('\n');
}

export function generateRuby(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = ['require "net/http"', 'require "uri"', ''];
  lines.push(`uri = URI("${url}")`);
  lines.push(`http = Net::HTTP.new(uri.host, uri.port)`);
  lines.push('http.use_ssl = true if uri.scheme == "https"');
  lines.push('');
  lines.push(`request = Net::HTTP::${method.charAt(0) + method.slice(1).toLowerCase()}.new(uri)`);
  for (const [k, v] of headersToEntries(headers)) {
    lines.push(`request["${k}"] = "${v}"`);
  }
  if (body) {
    lines.push(`request.body = '${body.replace(/'/g, "\\'")}'`);
  }
  lines.push('');
  lines.push('response = http.request(request)');
  lines.push('puts response.code');
  lines.push('puts response.body');
  return lines.join('\n');
}

export function generateRust(input: CodeGenInput): string {
  const { method, url, headers, body } = input;
  const lines: string[] = [
    'use reqwest;',
    '',
    '#[tokio::main]',
    'async fn main() -> Result<(), reqwest::Error> {',
    '    let client = reqwest::Client::new();',
  ];
  lines.push(`    let response = client.${method.toLowerCase()}("${url}")`);
  for (const [k, v] of headersToEntries(headers)) {
    lines.push(`        .header("${k}", "${v}")`);
  }
  if (body) {
    lines.push(`        .body(r#"${body}"#)`);
  }
  lines.push('        .send()');
  lines.push('        .await?;');
  lines.push('');
  lines.push('    println!("{}", response.status());');
  lines.push('    let text = response.text().await?;');
  lines.push('    println!("{}", text);');
  lines.push('    Ok(())');
  lines.push('}');
  return lines.join('\n');
}

export function generateCode(generatorId: string, input: CodeGenInput): string {
  switch (generatorId) {
    case 'curl': return generateCurl(input);
    case 'python-requests': return generatePythonRequests(input);
    case 'javascript-fetch': return generateJavaScriptFetch(input);
    case 'javascript-axios': return generateJavaScriptAxios(input);
    case 'nodejs-fetch': return generateNodeFetch(input);
    case 'dart-http': return generateDartHttp(input);
    case 'go-http': return generateGo(input);
    case 'java-http': return generateJava(input);
    case 'kotlin-okhttp': return generateKotlin(input);
    case 'php-curl': return generatePhp(input);
    case 'ruby-net': return generateRuby(input);
    case 'rust-reqwest': return generateRust(input);
    default: return generateCurl(input);
  }
}

File: packages/apidash_mcp_core/lib/src/codegen.dart
class CodeGenInput {
  final String method;
  final String url;
  final Map<String, String>? headers;
  final String? body;
  final String? contentType;

  CodeGenInput({
    required this.method,
    required this.url,
    this.headers,
    this.body,
    this.contentType,
  });
}

const supportedGenerators = [
  "curl", "python-requests", "javascript-fetch", "javascript-axios",
  "nodejs-fetch", "dart-http", "go-http", "java-http",
  "kotlin-okhttp", "php-curl", "ruby-net", "rust-reqwest"
];

String _sanitizeUrl(String url) {
  return url.replaceAll("'", "\\'");
}

String generateCurl(CodeGenInput input) {
  final lines = ["curl -X ${input.method} '${_sanitizeUrl(input.url)}'"];
  input.headers?.forEach((k, v) {
    lines.add("  -H '$k: $v'");
  });
  if (input.body != null && input.body!.isNotEmpty) {
    lines.add("  --data '${input.body!.replaceAll("'", "\\'")}'");
  }
  return lines.join(" \\\n");
}

String generatePythonRequests(CodeGenInput input) {
  final lines = ["import requests", "", "url = \"${input.url}\""];
  if (input.headers != null && input.headers!.isNotEmpty) {
    lines.add("headers = {");
    input.headers!.forEach((k, v) {
      lines.add("    \"$k\": \"$v\",");
    });
    lines.add("}");
  } else {
    lines.add("headers = {}");
  }

  if (input.body != null && input.body!.isNotEmpty) {
    lines.addAll([
      "",
      "payload = '''${input.body}'''",
      "",
      "response = requests.${input.method.toLowerCase()}(url, headers=headers, data=payload)"
    ]);
  } else {
    lines.addAll([
      "",
      "response = requests.${input.method.toLowerCase()}(url, headers=headers)"
    ]);
  }
  lines.addAll([
    "print(response.status_code)",
    "print(response.json())"
  ]);
  return lines.join("\n");
}

String generateJavaScriptFetch(CodeGenInput input) {
  final lines = ["const response = await fetch(\"${input.url}\", {", "  method: \"${input.method}\","];
  if (input.headers != null && input.headers!.isNotEmpty) {
    lines.add("  headers: {");
    input.headers!.forEach((k, v) {
      lines.add("    \"$k\": \"$v\",");
    });
    lines.add("  },");
  }
  if (input.body != null && input.body!.isNotEmpty) {
    final escaped = input.body!.replaceAll('\\', '\\\\').replaceAll('`', '\\`').replaceAll('\${', '\\\${');
    lines.add("  body: `$escaped`,");
  }
  lines.addAll([
    "});",
    "",
    "const data = await response.json();",
    "console.log(data);"
  ]);
  return lines.join("\n");
}

String generateJavaScriptAxios(CodeGenInput input) {
  final lines = ["import axios from \"axios\";", "", "const config = {", "  method: \"${input.method.toLowerCase()}\",", "  url: \"${input.url}\","];
  if (input.headers != null && input.headers!.isNotEmpty) {
    lines.add("  headers: {");
    input.headers!.forEach((k, v) {
      lines.add("    \"$k\": \"$v\",");
    });
    lines.add("  },");
  }
  if (input.body != null && input.body!.isNotEmpty) {
    lines.add("  data: ${input.body},");
  }
  lines.addAll([
    "};",
    "",
    "const { data } = await axios(config);",
    "console.log(data);"
  ]);
  return lines.join("\n");
}

String generateNodeFetch(CodeGenInput input) {
  final lines = ["const fetch = require(\"node-fetch\");", "", "async function run() {", "  const response = await fetch(\"${input.url}\", {", "    method: \"${input.method}\","];
  if (input.headers != null && input.headers!.isNotEmpty) {
    lines.add("    headers: {");
    input.headers!.forEach((k, v) {
      lines.add("      \"$k\": \"$v\",");
    });
    lines.add("    },");
  }
  if (input.body != null && input.body!.isNotEmpty) {
    lines.add("    body: JSON.stringify(${input.body}),");
  }
  lines.addAll([
    "  });",
    "  const data = await response.json();",
    "  console.log(data);",
    "}",
    "run();"
  ]);
  return lines.join("\n");
}

String generateDartHttp(CodeGenInput input) {
  final lines = ["import 'package:http/http.dart' as http;", "", "void main() async {", "  final uri = Uri.parse('${input.url}');"];
  bool hasHeaders = input.headers != null && input.headers!.isNotEmpty;
  if (hasHeaders) {
    lines.add("  final headers = {");
    input.headers!.forEach((k, v) {
      lines.add("    '$k': '$v',");
    });
    lines.add("  };");
  }
  if (input.body != null && input.body!.isNotEmpty) {
    lines.addAll([
      "  final body = '''${input.body}''';",
      "  final response = await http.${input.method.toLowerCase()}(uri, headers: ${hasHeaders ? 'headers' : '{}'}, body: body);"
    ]);
  } else {
    lines.add("  final response = await http.${input.method.toLowerCase()}(uri${hasHeaders ? ', headers: headers' : ''});");
  }
  lines.addAll([
    "  print(response.statusCode);",
    "  print(response.body);",
    "}"
  ]);
  return lines.join("\n");
}

String generateGo(CodeGenInput input) {
  final lines = [
    "package main", "",
    "import (", "  \"fmt\"", "  \"io\"", "  \"net/http\"", "  \"strings\"", ")", "",
    "func main() {"
  ];
  if (input.body != null && input.body!.isNotEmpty) {
    lines.addAll([
      "  payload := strings.NewReader(`${input.body}`)",
      "  req, _ := http.NewRequest(\"${input.method}\", \"${input.url}\", payload)"
    ]);
  } else {
    lines.add("  req, _ := http.NewRequest(\"${input.method}\", \"${input.url}\", nil)");
  }
  input.headers?.forEach((k, v) {
    lines.add("  req.Header.Add(\"$k\", \"$v\")");
  });
  lines.addAll([
    "  client := &http.Client{}",
    "  resp, _ := client.Do(req)",
    "  defer resp.Body.Close()",
    "  body, _ := io.ReadAll(resp.Body)",
    "  fmt.Println(resp.StatusCode)",
    "  fmt.Println(string(body))",
    "}"
  ]);
  return lines.join("\n");
}

String generateJava(CodeGenInput input) {
  final lines = [
    "import java.net.URI;",
    "import java.net.http.HttpClient;",
    "import java.net.http.HttpRequest;",
    "import java.net.http.HttpResponse;",
    "",
    "public class ApiRequest {",
    "  public static void main(String[] args) throws Exception {",
    "    var client = HttpClient.newHttpClient();",
    "    var request = HttpRequest.newBuilder()",
    "        .uri(URI.create(\"${input.url}\"))"
  ];
  input.headers?.forEach((k, v) {
    lines.add("        .header(\"$k\", \"$v\")");
  });
  if (input.body != null && input.body!.isNotEmpty) {
    final method = input.method;
    lines.add("        .$method(HttpRequest.BodyPublishers.ofString(\"\"\"${input.body}\"\"\"))");
  } else {
    lines.add("        .${input.method}(HttpRequest.BodyPublishers.noBody())");
  }
  lines.addAll([
    "        .build();",
    "    var response = client.send(request, HttpResponse.BodyHandlers.ofString());",
    "    System.out.println(response.statusCode());",
    "    System.out.println(response.body());",
    "  }",
    "}"
  ]);
  return lines.join("\n");
}

String generateKotlin(CodeGenInput input) {
  final lines = ["import okhttp3.OkHttpClient", "import okhttp3.Request", ""];
  if (input.body != null && input.body!.isNotEmpty) {
    lines.addAll([
      "import okhttp3.RequestBody.Companion.toRequestBody",
      "import okhttp3.MediaType.Companion.toMediaType",
      ""
    ]);
  }
  lines.add("val client = OkHttpClient()");
  if (input.body != null && input.body!.isNotEmpty) {
    lines.add("val body = \"\"\"${input.body}\"\"\".toRequestBody(\"application/json\".toMediaType())");
  }
  lines.addAll([
    "val request = Request.Builder()",
    "    .url(\"${input.url}\")"
  ]);
  input.headers?.forEach((k, v) {
    lines.add("    .addHeader(\"$k\", \"$v\")");
  });
  if (input.body != null && input.body!.isNotEmpty) {
    lines.add("    .${input.method.toLowerCase()}(body)");
  } else {
    lines.add("    .${input.method.toLowerCase()}()");
  }
  lines.addAll([
    "    .build()",
    "",
    "val response = client.newCall(request).execute()",
    "println(response.code)",
    "println(response.body?.string())"
  ]);
  return lines.join("\n");
}

String generatePhp(CodeGenInput input) {
  final lines = ["<?php", "", "\$ch = curl_init(\"${input.url}\");", "curl_setopt(\$ch, CURLOPT_RETURNTRANSFER, true);", "curl_setopt(\$ch, CURLOPT_CUSTOMREQUEST, \"${input.method}\");"];
  if (input.headers != null && input.headers!.isNotEmpty) {
    lines.add("curl_setopt(\$ch, CURLOPT_HTTPHEADER, [");
    input.headers!.forEach((k, v) {
      lines.add("  \"$k: $v\",");
    });
    lines.add("]);");
  }
  if (input.body != null && input.body!.isNotEmpty) {
    lines.add("curl_setopt(\$ch, CURLOPT_POSTFIELDS, '${input.body!.replaceAll("'", "\\'")}');");
  }
  lines.addAll([
    "",
    "\$response = curl_exec(\$ch);",
    "\$statusCode = curl_getinfo(\$ch, CURLINFO_HTTP_CODE);",
    "curl_close(\$ch);",
    "",
    "echo \$statusCode . \"\\n\";",
    "echo \$response . \"\\n\";"
  ]);
  return lines.join("\n");
}

String generateRuby(CodeGenInput input) {
  final lines = ["require \"net/http\"", "require \"uri\"", "", "uri = URI(\"${input.url}\")", "http = Net::HTTP.new(uri.host, uri.port)", "http.use_ssl = true if uri.scheme == \"https\"", ""];
  final m = input.method;
  final methodClass = m.substring(0, 1).toUpperCase() + m.substring(1).toLowerCase();
  lines.add("request = Net::HTTP::$methodClass.new(uri)");
  input.headers?.forEach((k, v) {
    lines.add("request[\"$k\"] = \"$v\"");
  });
  if (input.body != null && input.body!.isNotEmpty) {
    lines.add("request.body = '${input.body!.replaceAll("'", "\\'")}'");
  }
  lines.addAll([
    "",
    "response = http.request(request)",
    "puts response.code",
    "puts response.body"
  ]);
  return lines.join("\n");
}

String generateRust(CodeGenInput input) {
  final lines = [
    "use reqwest;",
    "",
    "#[tokio::main]",
    "async fn main() -> Result<(), reqwest::Error> {",
    "    let client = reqwest::Client::new();",
    "    let response = client.${input.method.toLowerCase()}(\"${input.url}\")"
  ];
  input.headers?.forEach((k, v) {
    lines.add("        .header(\"$k\", \"$v\")");
  });
  if (input.body != null && input.body!.isNotEmpty) {
    lines.add("        .body(r#\"${input.body}\"#)");
  }
  lines.addAll([
    "        .send()",
    "        .await?;",
    "",
    "    println!(\"{}\", response.status());",
    "    let text = response.text().await?;",
    "    println!(\"{}\", text);",
    "    Ok(())",
    "}"
  ]);
  return lines.join("\n");
}

String generateCode(String generatorId, CodeGenInput input) {
  switch (generatorId) {
    case 'curl': return generateCurl(input);
    case 'python-requests': return generatePythonRequests(input);
    case 'javascript-fetch': return generateJavaScriptFetch(input);
    case 'javascript-axios': return generateJavaScriptAxios(input);
    case 'nodejs-fetch': return generateNodeFetch(input);
    case 'dart-http': return generateDartHttp(input);
    case 'go-http': return generateGo(input);
    case 'java-http': return generateJava(input);
    case 'kotlin-okhttp': return generateKotlin(input);
    case 'php-curl': return generatePhp(input);
    case 'ruby-net': return generateRuby(input);
    case 'rust-reqwest': return generateRust(input);
    default: return generateCurl(input);
  }
}
